Add javaFxPart5 guide

flexibo opened the pull request.

damithc commented on tutorials/javaFxPart5.md:
````diff
@@ -0,0 +1,419 @@
+Update the `start` method as shown below.
+
+**Main.java**
+```java
````
```suggestion
```java {heading="Main.java" highlight-lines="4-5"}
```

Just a suggestion only.

flexibo replied:
Currently the highlighted lines look something like this:

![image](https://github.com/se-edu/guides/assets/40513197/a48b6866-aacb-4e3c-9aa0-b0cb744e4f11)

Is there a way to make the highlighted portions more brighter/more distinct? Can't seem to figure out from the MarkBind guide.

damithc replied:
@flexibo MarkBind does not have a way to change the code highlight color.

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+
+# Customize with JavaFX CSS!
+
+For current and future projects, you might need to customize colors, shapes, borders, and other design elements.
```
```suggestion
Suppose you want to customize colors, shapes, borders, and other design elements.
```

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+For current and future projects, you might need to customize colors, shapes, borders, and other design elements.
+The most effective way to do this is through CSS, so this will be a mini [**JavaFX CSS**](https://docs.oracle.com/javafx/2/api/javafx/scene/doc-files/cssref.html)
+introductory guide.
+
```
```suggestion
<box type="tip" seamless>

If you are new to CSS entirely, a quick way to learn the basics is [give a suitable resoruce here].
</box>
```

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+- Run the app
+- Select both image and label to view the distance between them
+  ![View effects of margin](images/javafx/ViewEffectsOfMargin.png)
+
```
Should we also include how to fix the text clipping problem? https://github.com/nus-cs2103-AY2324S2/forum/issues/208

flexibo replied:
Yup, sounds good, will work on this.

flexibo replied:
@damithc to follow up on this, since the issue is fixed by removing the `prefHeight` for VBox, is it possible to adjust that in the original code from the earlier parts instead?

damithc replied:
Sure, that works too.

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+
+.button {
+    -fx-background-color: mediumspringgreen;
+    -fx-font: italic bold 16px "Arial";
```
Using a fixed-width font helps with the alignments. Compare the two below:

[x] First item
[ ] Second item

vs
```
[x] First item
[ ] Second item
```

flexibo replied:
@damithc the varying font size is applied to the send button only, the checks are unaffected (just needed to make adjustments to number of space when a task is not marked):

![image](https://github.com/se-edu/guides/assets/40513197/a44eed57-2ae6-4d48-bafb-eb7aa19e5d32)

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+
+The values are set in a **clockwise** manner starting with the **top left corner/top edge**.
+
+![BorderDemo](images/javafx/BorderDemo.png)
```
```suggestion
<pic src="images/javafx/BorderDemo.png" width="600" />
```

Otherwise the image appears disproportionate to the rest of the document.

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+## Color
+
+There are a few ways to set a color.
+1. Using `rgb(...)` / `rgba(...)` (use the fourth parameter to set opacity from 0 to 1)
```
Give an example, especially for the opacity?

damithc commented on tutorials/javaFxPart5.md:
````diff
@@ -0,0 +1,419 @@
+-fx-background-color: linear-gradient(to bottom right, red, blue);
+-fx-background-color: radial-gradient(radius 100%, red, purple, blue);
+```
+
````
Also show the result, using a screenshot?

damithc commented on tutorials/javaFxPart5.md:
````diff
@@ -0,0 +1,419 @@
+    ...
+    -fx-background-color: duke;   /* Using looked-up color */
+}
+```
````
Use a meaningful name for the color?

flexibo replied:
Hmmm, can't really think of an appropriate name, `themecolor` maybe?

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+the shadow effect will not apply to the transparent areas.
+
+![Transparent background image demo](images/javafx/ImageShadowDemo.png)
+
```
Using speech bubble as a caption might cause confusion. Perhaps crop it out and use something like this instead?

```suggestion
Drop shadow | Inner shadow | Transparent background
------------|--------------|-----------------------
![](images/javafx/ImageShadowDemo.png)| ![](images/javafx/ImageShadowDemo.png) | ![](images/javafx/ImageShadowDemo.png)
```suggestion

```

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+- [Apply style to just that specific node with the id (e.g., `#displayPicture`) instead](#image-and-shadows)
+
+
+# Other things to consider
```
For each of the items below, can we give a hint or a link to a suitable resource?

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+- Set GIF as profile pictures
+- Change the text box color if it is an error message
+- Add [sound effects](https://docs.oracle.com/javase/8/javafx/api/javafx/scene/media/AudioClip.html)
+  when sending messages
```
How about setting a background image for the main chat area (and reducing opacity so that it doesn't  clash too much with the chat bubbles)?

flexibo replied:
Sounds good, should I make a guide on this or do I add that under "Other things to consider"?

damithc replied:
@flexibo My personal preference is to include it in the main part of this guide, but I'll leave it to you.

flexibo replied:
@damithc Noted, is there anything else in the list that you think it is more suitable to be part of the main guide instead?

damithc replied:
> @damithc Noted, is there anything else in the list that you think it is more suitable to be part of the main guide instead?

If not covered already, it might be good to explain (in the main body) a tweak that changes the UI dynamically based on the command typed e.g., change the speech bubble style if there is an error, or vary the speech bubble style based on the command.

damithc commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+
+--------------------------------------------------------------------------------
+**Authors:**
+* Initial Version: Debbie Hii 
```
```suggestion
* Initial Version: Debbie Hii (@flexibo)
```

damithc reviewed (commented):
Nice work @flexibo
I've added some comments.

flexibo commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+Open `src/main/resources/view/MainWindow.fxml` in SceneBuilder and click on `TextField`. On the right accordion pane you
+should be able to see the AnchorPane constraints of the `TextField`.
+
+![SceneBuilder opening MainWindow.fxml with Layout view](images/javafx/TextFieldAnchorPaneConstraints1.png)
```
Will look into this, thanks!

flexibo commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+Open `src/main/resources/view/MainWindow.fxml` in SceneBuilder and click on `TextField`. On the right accordion pane you
+should be able to see the AnchorPane constraints of the `TextField`.
+
+![SceneBuilder opening MainWindow.fxml with Layout view](images/javafx/TextFieldAnchorPaneConstraints1.png)
```
Sure thing, is it sufficient to just provide the fxml code? Or will it be better to explain the code as well?

flexibo commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+- Anchor the `ScrollPane` to resize both horizontally and vertically
+
+### Automatically Anchor TextField
+Open `src/main/resources/view/MainWindow.fxml` in SceneBuilder and click on `TextField`. On the right accordion pane you
+should be able to see the AnchorPane constraints of the `TextField`.
```
Noted, thanks!

flexibo commented on tutorials/javaFxPart5.md:
```diff
@@ -0,0 +1,419 @@
+
+![Update AnchorPane constraints of ScrollPane](images/javafx/ScrollPaneAnchorPaneConstraints.png)
+
+Save `MainWindow.fxml` and you're done! Run and test your app out to see if it works when you resize the window.
```
AFAIK, if you anchor only the `TextField`, the `ScrollPane` wont be resized. Likewise if the `ScrollPane` gets anchored, the `TextField` wont be resized unless I anchor both elements. Unless I misunderstood this.

damithc replied:
I see. In that case, perhaps suggest the reader to try resizing after anchoring the first element, and explain why it doesn't work fully yet? It might help the reader to better understand how the anchoring mechanism works.

flexibo replied:
Will do!

damithc commented:
@flexibo Let me know when the PR is ready for another review.

flexibo commented:
@damithc Hi prof, here are some of the general changes I've made aside from the ones that you've requested:
- Update the diagrams to make it look more professional (including videos), unsure if I added too much videos
- Add more elaborations and examples for customizing color CSS 
- Adjusted the side nav to see H3s as well
- Add `## Change style dynamically based on command`, but I'm unsure if the instructions are clear enough as some of the code may differ for students. I'm also unsure if the current example code violates any SWE principles. 
- Add `## Set Image as background`

damithc commented:
Great work @flexibo 💯 
I've merged it for now, so that it can appear in the website already. Minor tweaks can be done as a separate PR later.
